crates/bevy_math/src/curves/variable_linear.rs

```diff
@@ -0,0 +1,243 @@
+use crate::{
+    curves::{Curve, CurveCursor, CurveError},
+    interpolation::Lerp,
+};
+
+// TODO: Curve/Clip need a validation during deserialization because they are
+// structured as SOA (struct of arrays), so the vec's length must match
+
+// TODO: impl Serialize, Deserialize
+/// Curve with sparse keyframes frames, in another words a curve with variable frame rate;
+///
+/// This is a very useful curve, because it can accommodate the output of a linear reduction keyframe algorithm
+/// to lower the memory foot print. As a down side it requires the use of a keyframe cursor, and
+/// loses performance when the curve frame rate is higher than the curve sampling frame rate;
+///
+/// It can't handle discontinuities, as in two keyframes with the same timestamp.
+///
+/// **NOTE** Keyframes count is limited by the [`CurveCursor`] size.
+#[derive(Default, Debug)]
+pub struct CurveVariableLinear<T> {
+    time_stamps: Vec<f32>,
+    keyframes: Vec<T>,
+}
+
+impl<T: Clone> Clone for CurveVariableLinear<T> {
+    fn clone(&self) -> Self {
+        Self {
+            time_stamps: self.time_stamps.clone(),
+            keyframes: self.keyframes.clone(),
+        }
+    }
+}
+
+impl<T> CurveVariableLinear<T> {
+    pub fn new(samples: Vec<f32>, values: Vec<T>) -> Result<Self, CurveError> {
+        let length = samples.len();
+
+        // Make sure both have the same length
+        if length != values.len() {
+            return Err(CurveError::MismatchedLength);
+        }
+
+        if values.len() > CurveCursor::MAX as usize {
+            return Err(CurveError::KeyframeLimitReached(CurveCursor::MAX as usize));
+        }
+
+        // Make sure the
+        if !samples
+            .iter()
+            .zip(samples.iter().skip(1))
+            .all(|(a, b)| a < b)
+        {
+            return Err(CurveError::NotSorted);
+        }
+
+        Ok(Self {
+            time_stamps: samples,
+            keyframes: values,
+        })
+    }
+
+    pub fn from_line(t0: f32, t1: f32, v0: T, v1: T) -> Self {
+        if t0 < t1 {
+            Self {
+                time_stamps: vec![t0, t1],
+                keyframes: vec![v0, v1],
+            }
+        } else {
+            Self {
+                time_stamps: vec![t1, t0],
+                keyframes: vec![v1, v0],
+            }
+        }
+    }
+
+    pub fn from_constant(v: T) -> Self {
+        Self {
+            time_stamps: vec![0.0],
+            keyframes: vec![v],
+        }
+    }
+
+    pub fn insert(&mut self, time: f32, value: T) {
+        // Keyframe length is limited by the cursor size yype that is 2 bytes,
+        assert!(
```

Comment:
ditto on documenting the `panic!` behavior.

Reply:
done